File: TopCoder/src/main/java/com/leetcode/array/leetcode/MinPathSum.java
package com.leetcode.array.leetcode;

//https://leetcode.com/problems/minimum-path-sum/description/
public class MinPathSum {

    public void find (int num[][]) {
        int rowLength = num.length;
        int colLength = num[0].length;
        int dp[] = new int[colLength];
        dp[0] = num[0][0];
        for ( int i = 1 ; i < colLength ; i++) {
            dp[i] = num[0][i] + dp[i-1];
        }

        for (int i = 1 ; i < rowLength ; i++) {
            for (int j = 0 ; j < colLength ; j++) {
               if ( j == 0) {
                   dp[j] = dp[j] + num[i][j];
               } else {
                   dp[j] = num[i][j] + Math.min(dp[j], dp[j-1]);
               }
            }
        }
        System.out.println(dp[colLength-1]);

    }

    public static void main(String[] args) {
        int[][] matrix = {
                {1,3,1},
                {1,5,1},
                {4,2,1}
        };
        new MinPathSum().find(matrix);
    }
}

File: TopCoder/python/MinPathSum.py
from typing import List

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        colLength = len(grid[0])
        rowLength = len(grid)
        dp = [[0]*colLength for _ in range(rowLength)]

        for i in range(0, rowLength):
            for j in range(0, colLength):
                if i == 0 and j == 0:
                    dp[i][j] = grid[i][j]
                elif i == 0:
                    dp[i][j] = grid[i][j] + dp[i][j-1]
                elif j == 0:
                    dp[i][j] = grid[i][j] + dp[i-1][j]
                else:
                    dp[i][j] = grid[i][j] + min(dp[i-1][j], dp[i][j-1])

        return dp[rowLength-1][colLength-1]

if __name__ == "__main__":
   matrix = [
       [1,3,1],
       [1,5,1],
       [4,2,1]
   ]
   sol = Solution()
   res = sol.minPathSum(matrix)
   print(res)